<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Curriculum Vitae | Nhi V. Vo </title> <meta name="author" content="Nhi V. Vo"> <meta name="description" content=""> <meta name="keywords" content="academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9C%B7&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://nhinvo.github.io/cv/"> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Nhi</span> V. Vo </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">Home </a> </li> <li class="nav-item active"> <a class="nav-link" href="/cv/">Curriculum Vitae <span class="sr-only">(current)</span> </a> </li> <li class="nav-item "> <a class="nav-link" href="/repositories/">Github Repo </a> </li> <li class="nav-item "> <a class="nav-link" href="/interests/">Interests </a> </li> <li class="nav-item "> <a class="nav-link" href="/blog/">Blogs </a> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main" style="max-width: 930px;"> <div class="post"> <header class="post-header"> <h1 class="post-title"> Curriculum Vitae </h1> </header> <article> <div class="cv"> <a class="anchor" id="Experience"></a> <div class="card mt-3 p-3"> <h3 class="card-title" style="font-size: 2rem;">Experience</h3> <div> <ul class="card-text font-weight-light list-group list-group-flush"> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">Feb 23 - Present</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> Cambridge, MA </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Bioinformatics Research Associate II</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Massachusetts Institute of Technology - The Chisolm Lab</td> </tr> </tbody> </table> </div> </div> </li> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">May 22 - Aug 22</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> Gainesville, FL </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Research Assistant</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Florida Museum of Natural History - The Kawahara Lab</td> </tr> </tbody> </table> </div> </div> </li> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">May 22 - Aug 22</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> West Lafayette, IN </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Assistant Data Analyst</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Purdue University - Fernandez-Juricic Lab</td> </tr> </tbody> </table> </div> </div> </li> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">Aug 21 - May 22</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> West Lafayette, IN </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Research Assistant</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Purdue University - Kihara Bioinformatics Laboratory</td> </tr> </tbody> </table> </div> </div> </li> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">Aug 21 - Dec 22</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> West Lafayette, IN </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Technology Supervisor</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Information Technology at Purdue (ITaP)</td> </tr> </tbody> </table> </div> </div> </li> </ul> </div> </div> <a class="anchor" id="Publications"></a> <div class="card mt-3 p-3"> <h3 class="card-title" style="font-size: 2rem;">Publications</h3> <div> <div class="publications"> <ol class="bibliography"> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100">Accepted</abbr> <figure> <picture> <img src="/assets/img/publication_preview/ProSynTax.gif" class="preview z-depth-1 rounded" width="100%" height="auto" alt="ProSynTax.gif" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="Coe2025.03.20.644373" class="col-sm-8"> <div class="title">ProSynTax: A curated protein dataset for taxonomic classification of Prochlorococcus and Synechococcus in metagenomes</div> <div class="author"> Allison Coe<sup>*</sup>, James I. Mullet<sup>*</sup>, <em>Nhi N. Vo<sup>*</sup></em>, Paul M. Berube, Maya I. Anjur-Dietrich, Eli Salcedo, and <span class="more-authors" title="click to view 7 more authors" onclick=" var element=$(this); element.attr('title', ''); var more_authors_text=element.text() == '7 more authors' ? 'Sierra M. Parker, Konnor VonEmster, Christina Bliem, Aldo A. Arellano, Kurt G. Castro, Jamie W. Becker, Sallie W. Chisholm' : '7 more authors'; var cursorPosition=0; var textAdder=setInterval(function(){ element.html(more_authors_text.substring(0, cursorPosition + 1)); if (++cursorPosition == more_authors_text.length){ clearInterval(textAdder); } }, '1'); ">7 more authors</span> </div> <div class="periodical"> Oct 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://doi.org/10.1101/2025.03.20.644373" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> <a href="https://github.com/jamesm224/ProSynTax-workflow" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>Prochlorococcus and Synechococcus are abundant marine picocyanobacteria that contribute significantly to ocean primary production. Recent genome sequencing efforts, including those presented here, have yielded a large number of high-quality reference genomes, enabling the classification of these picocyanobacteria in marine metagenomic sequence data at high phylogenetic resolution. When combined with environmental data, these classifications can guide cluster/clade/grade assignments and offer insights into niche differentiation within these populations. Here we present ProSynTax, a curated protein sequence dataset and accompanying workflow aimed at enhancing the taxonomic resolution of Prochlorococcus and Synechococcus classification. ProSynTax includes proteins from 1,260 genomes of Prochlorococcus and Synechococcus, including single-amplified genomes, high-quality draft genomes, and newly closed genomes. Additionally, ProSynTax incorporates proteins from 41,753 genomes of marine heterotrophic bacteria, archaea, and viruses to assess microbial and viral communities surrounding Prochlorococcus and Synechococcus. This resource enables accurate classification of picocyanobacterial clusters/clades/grades in metagenomic data – even when present at 0.15% of reads for Prochlorococcus or 0.03% of reads for Synechococcus.Competing Interest StatementThe authors have declared no competing interest.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100">ISME Communications</abbr> <figure> <picture> <img src="/assets/img/publication_preview/dark-rnaseq.gif" class="preview z-depth-1 rounded" width="100%" height="auto" alt="dark-rnaseq.gif" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="10.1093/ismeco/ycae131" class="col-sm-8"> <div class="title">Emergence of metabolic coupling to the heterotroph Alteromonas promotes dark survival in Prochlorococcus</div> <div class="author"> Allison Coe, Rogier Braakman, Steven J Biller, Aldo Arellano, Christina Bliem, <em>Nhi N. Vo</em>, and <span class="more-authors" title="click to view 6 more authors" onclick=" var element=$(this); element.attr('title', ''); var more_authors_text=element.text() == '6 more authors' ? 'Konnor Emster, Elaina Thomas, Michelle DeMers, Claudia Steglich, Jef Huisman, Sallie W Chisholm' : '6 more authors'; var cursorPosition=0; var textAdder=setInterval(function(){ element.html(more_authors_text.substring(0, cursorPosition + 1)); if (++cursorPosition == more_authors_text.length){ clearInterval(textAdder); } }, '1'); ">6 more authors</span> </div> <div class="periodical"> <em>ISME Communications</em>, Oct 2024 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://doi.org/10.1093/ismeco/ycae131" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> </div> <div class="abstract hidden"> <p>Prochlorococcus is found throughout the euphotic zone in the oligotrophic open ocean. Deep mixing and sinking while attached to particles can, however, transport Prochlorococcus cells below this sunlit zone, depriving them of light for extended periods of time. Previous work has shown that Prochlorococcus by itself cannot survive extended periods of darkness. However, when co-cultured with a heterotrophic microbe and subjected to repeated periods of extended darkness, Prochlorococcus cells develop an epigenetically inherited dark-tolerant phenotype that can survive longer periods of darkness. Here we examine the metabolic and physiological changes underlying this adaptation using co-cultures of dark-tolerant and parental strains of Prochlorococcus, each grown with the heterotroph Alteromonas under diel light:dark conditions. The relative abundance of Alteromonas was higher in dark-tolerant than parental co-cultures, while dark-tolerant Prochlorococcus cells were larger, contained less chlorophyll, and were less synchronized to the light:dark cycle. Meta-transcriptome analysis revealed that dark-tolerant co-cultures undergo a joint change, in which Prochlorococcus undergoes a relative shift from photosynthesis to respiration, while Alteromonas shifts toward using more organic acids instead of sugars. Furthermore, the transcriptome data suggested enhanced biosynthesis of amino acids and purines in dark-tolerant Prochlorococcus and enhanced degradation of these compounds in Alteromonas. Collectively, our results demonstrate that dark adaptation involves a strengthening of the metabolic coupling between Prochlorococcus and Alteromonas, presumably mediated by an enhanced, and compositionally modified, carbon exchange between the two species.</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100">Under Review</abbr> <figure> <picture> <img src="/assets/img/publication_preview/pro-het-phage-interaction.gif" class="preview z-depth-1 rounded" width="100%" height="auto" alt="pro-het-phage-interaction.gif" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="Coe2025.09.17.676900" class="col-sm-8"> <div class="title">Influence of heterotrophs on phage infection of marine picocyanobacteria</div> <div class="author"> Allison Coe, Sierra M. Parker, <em>Nhi N. Vo</em>, Sean M. Kearney, Shaul Pollak, Konnor Emster, and <span class="more-authors" title="click to view 3 more authors" onclick=" var element=$(this); element.attr('title', ''); var more_authors_text=element.text() == '3 more authors' ? 'James I. Mullet, Kurt G. Castro, Sallie W. Chisholm' : '3 more authors'; var cursorPosition=0; var textAdder=setInterval(function(){ element.html(more_authors_text.substring(0, cursorPosition + 1)); if (++cursorPosition == more_authors_text.length){ clearInterval(textAdder); } }, '1'); ">3 more authors</span> </div> <div class="periodical"> Oct 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://doi.org/10.1101/2025.09.17.676900" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> <a href="https://github.com/nhinvo/pro-het-phage-interaction" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>Picocyanobacteria Prochlorococcus and Synechococcus coexist with both their lytic phages and heterotrophic bacteria in the oceans. These lytic phages are a significant cause of mortality, and heterotrophic bacteria have been shown to increase the fitness of Prochlorococcus by reducing oxidative stress and cross feeding under extended darkness. Studies of Prochlorococcus-phage interactions are often done with xenic cultures as it has been historically difficult to obtain and maintain heterotroph-free cultures. Here we examine the effects of heterotrophic bacteria on phage infection in Prochlorococcus and Synechococcus by comparing phage infection dynamics in cultures with and without heterotrophs present. We found that Prochlorococcus populations resumed growth following infection only in the presence of heterotrophs, independent of phage:host or heterotroph:host ratios. In phage:host pairing with Synechococcus the outcomes varied, suggesting that the impact of heterotrophs on phage infection may be dependent on the phage:host interaction. In cases where the host recovered from phage infection, heterotrophs appeared to facilitate it both by mitigating oxidative stress and possibly supplying organic carbon sources, which may support post-infection growth. Furthermore, Prochlorococcus and Synechococcus populations that recovered from infection were resistant to phage infection when transferred to fresh media. Evidence argues against genetic change as the mechanism of resistance, suggesting that Prochlorococcus and Synechococcus populations in co-culture with heterotrophs undergo non-genetic adaptations during recovery from phage infection, likely driven by heterotroph-derived organic compounds that reshape host metabolism and confer protection against future lysis.Competing Interest StatementThe authors have declared no competing interest.Simons Foundation, 337262, 329108, 649394</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100">Under Review</abbr> <figure> <picture> <img src="/assets/img/publication_preview/biofilm.gif" class="preview z-depth-1 rounded" width="100%" height="auto" alt="biofilm.gif" data-zoomable loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div id="Anjur-Dietrich2025.08.05.668435" class="col-sm-8"> <div class="title">Biofilm formation and dynamics in the marine cyanobacterium Prochlorococcus</div> <div class="author"> Maya I. Anjur-Dietrich, Katelyn G. Jones, James I. Mullet, <em>Nhi N. Vo</em>, Kurt G. Castro, Sierra M. Parker, and <span class="more-authors" title="click to view 1 more author" onclick=" var element=$(this); element.attr('title', ''); var more_authors_text=element.text() == '1 more author' ? 'Sallie W. Chisholm' : '1 more author'; var cursorPosition=0; var textAdder=setInterval(function(){ element.html(more_authors_text.substring(0, cursorPosition + 1)); if (++cursorPosition == more_authors_text.length){ clearInterval(textAdder); } }, '1'); ">1 more author</span> </div> <div class="periodical"> Oct 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://doi.org/10.1101/2025.08.05.668435" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> <a href="https://github.com/nhinvo/biofilm-prochlorococcus" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">Code</a> </div> <div class="abstract hidden"> <p>The picocyanobacterium Prochlorococcus is responsible for ~10% of annual marine carbon fixation and plays a role in the global carbon budget. While these phototrophs are primarily considered free-living and neutrally buoyant in the euphotic zone, we observe that they can form biofilms on diverse substrates. This trait is conserved across Prochlorococcus ecotypes, and populations continuously transition between planktonic and biofilm states via a non-genetic heritable mechanism. Throughout their growth, cells in biofilms retain a reversible, dynamic attachment state, and measurements of growth, photosynthesis, and respiration rates reveal that cells in biofilms exude more organic carbon than their planktonic counterparts. Estimates of the fraction of Prochlorococcus cells attached to particles in the ocean reveal that a significant adherent population exists throughout the euphotic and mesopelagic zones. This work describes a new dimension of Prochlorococcus’s ecological niche and suggests a role in carbon export to the deep sea.Competing Interest StatementThe authors have declared no competing interest.Simons Foundation, https://ror.org/01cmst727, 337262FY23, 721246, 984601</p> </div> </div> </div> </li> <li> <div class="row"> <div class="col col-sm-2 abbr"> <abbr class="badge rounded w-100">Submitted</abbr> </div> <div id="Ziegler2025.09.25.678681" class="col-sm-8"> <div class="title">The shared and distinct roles of Prochlorococcus and co-occurring heterotrophic bacteria in regulating community dynamics</div> <div class="author"> Christine A. Ziegler, James I. Mullet, Allison Coe, <em>Nhi N. Vo</em>, Eli Salcedo, Dillon M. Arrigan, and <span class="more-authors" title="click to view 2 more authors" onclick=" var element=$(this); element.attr('title', ''); var more_authors_text=element.text() == '2 more authors' ? 'Sierra M. Parker, Sallie W. Chisholm' : '2 more authors'; var cursorPosition=0; var textAdder=setInterval(function(){ element.html(more_authors_text.substring(0, cursorPosition + 1)); if (++cursorPosition == more_authors_text.length){ clearInterval(textAdder); } }, '1'); ">2 more authors</span> </div> <div class="periodical"> Oct 2025 </div> <div class="periodical"> </div> <div class="links"> <a class="abstract btn btn-sm z-depth-0" role="button">Abs</a> <a href="https://doi.org/10.1101/2025.09.25.678681" class="btn btn-sm z-depth-0" role="button" rel="external nofollow noopener" target="_blank">DOI</a> </div> <div class="abstract hidden"> <p>Prochlorococcus is the world’s most abundant photosynthetic organism with over 1027 cells distributed across much of Earth’s oceans, and is collectively responsible for almost 10% of marine carbon fixation. Naturally co-occurring heterotrophic bacteria at roughly 105-106 cells mL-1 in the oceans have been shown to increase Prochlorococcus fitness and productivity. Despite this massive scale, our understanding of these globally important interactions remains limited, with past research largely focused on single Prochlorococcus-heterotroph pairings involving only a few species. In this study, we extend this perspective by using synthetic communities containing multiple diverse heterotrophic strains isolated from Prochlorococcus enrichment cultures. Specifically, we isolated the four most abundant co-occurring heterotroph species and examined both individual Prochlorococcus–heterotroph interactions and interactions within a synthetic community comprising Prochlorococcus and all four heterotrophs. Using absolute quantification of RNA, DNA, and cell counts over the course of Prochlorococcus growth curves, we find that Prochlorococcus has a modest, species-independent transcriptional response to heterotrophs, whereas each heterotroph displays a markedly different transcriptional response to the community and fulfills distinct metabolic roles. Transcriptional analyses reveal several potential crossfeeding interactions and indicate that community dynamics are influenced not only by metabolic activity but also antagonistic mechanisms, defense responses, and coordinated group behaviors. By pairing synthetic community approaches with absolute abundance measurements, we can gain deeper insight into the forces that shape microbial community assembly in the oceans and their role in driving the global carbon cycle.Competing Interest StatementThe authors have declared no competing interest.Simons Foundation, https://ror.org/01cmst727, 337262, 721246, LS-FMME-00003951</p> </div> </div> </div> </li> </ol> </div> </div> </div> <a class="anchor" id="Education"></a> <div class="card mt-3 p-3"> <h3 class="card-title" style="font-size: 2rem;">Education</h3> <div> <ul class="card-text font-weight-light list-group list-group-flush"> <li class="list-group-item"> <div class="row"> <div class="col-xs-2 cl-sm-2 col-md-2 text-center" style="width: 75px"> <table class="table-cv"> <tbody> <tr> <td> <span class="badge font-weight-bold danger-color-dark text-uppercase align-middle" style="min-width: 75px">Aug 19 - Dec 22</span> </td> </tr> <tr> <td> <p class="location"> <i class="fa-solid fa-location-dot iconlocation"></i> West Lafayette, IN </p> </td> </tr> </tbody> </table> </div> <div class="col-xs-10 cl-sm-10 col-md-10 mt-2 mt-md-0"> <h6 class="title font-weight-bold ml-1 ml-md-4">Bachelor of Science in Genetics</h6> <table class="table-cv ml-1 ml-md-4 institution"> <tbody> <tr> <td style="vertical-align: top; text-align: center" class="institution"> <i class="fa-solid fa-building-columns iconinstitution"></i> </td> <td class="institution">Purdue University</td> </tr> </tbody> </table> <ul class="items"> <li> <span class="item">Minors: Bioinformatics, Chemistry</span> </li> <li> <span class="item">GPA: 3.88/4.00</span> </li> <li> <span class="item">Dean's List &amp; Semester Honors for 5 semesters</span> </li> </ul> </div> </div> </li> </ul> </div> </div> <a class="anchor" id="Technical Skills"></a> <div class="card mt-3 p-3"> <h3 class="card-title" style="font-size: 2rem;">Technical Skills</h3> <div> <ul class="card-text font-weight-light list-group list-group-flush"> <li class="list-group-item"> <b>Programming:</b> Python, Bash/Shell, Linux/Unix, SLURM/HPC, Git/GitHub, Jupyter Notebook, R.</li> <li class="list-group-item"> <b>Data Analysis &amp; Visualization:</b> Pandas, NumPy, SciPy, Matplotlib, Plotly, Seaborn, Adobe Illustrator, Excel, Scikit-Learn.</li> <li class="list-group-item"> <b>Workflow Languages:</b> Snakemake, Nextflow.</li> <li class="list-group-item"> <b>NGS Datasets:</b> Illumina, Oxford Nanopore, PacBio, Sanger, RNA-seq, (meta)transcriptomics, DNA-seq, (meta)genomics, epigenomics, single cell genomics.</li> <li class="list-group-item"> <b>Bioinformatics Analyses:</b> Whole Genome Assembly, Metagenomic Assembly (MAGs), Comparative Genomics Analysis, Pangenomics, Variant Calling, Differential Gene Expression Analysis, Taxonomic Profiling.</li> </ul> </div> </div> <a class="anchor" id="Other Interests"></a> <div class="card mt-3 p-3"> <h3 class="card-title" style="font-size: 2rem;">Other Interests</h3> <div> <ul class="card-text font-weight-light list-group list-group-flush"> <li class="list-group-item"> <h5 class="font-italic">Currently Reading - as of 11/17/25</h5> <ul class="subitems"> <li> <span class="subitem">System Collapse - Martha Wells (6th book in the Murderbot Diaries series)</span> </li> <li> <span class="subitem">The Two Towers - J.R.R Tolkien (2nd book in the Lord of The Rings series)</span> </li> </ul> </li> <li class="list-group-item"> <h5 class="font-italic">Music/Guitar</h5> <ul class="subitems"> <li> <span class="subitem">I play the guitar and I find tabbing songs by ear incredibly rewarding.</span> </li> <li> <span class="subitem">All my guitar tabs are accessible on my <a href="https://nhinvo.github.io/interests/#Guitar%20Tabs">Interest</a> page.</span> </li> <li> <span class="subitem">I am self-taught and am not musically trained therefore the tabs might not be perfect.</span> </li> </ul> </li> <li class="list-group-item"> <h5 class="font-italic">Arts &amp; Crafts</h5> <ul class="subitems"> <li> <span class="subitem">Check out my work <a href="https://nhinvo.github.io/projects/my-art/">here</a>!</span> </li> </ul> </li> <li class="list-group-item"> <h5 class="font-italic"></h5> <ul class="subitems"> <li> <span class="subitem">Always happy to discuss books, art and music, along with receiving your recommendations!</span> </li> </ul> </li> </ul> </div> </div> </div> </article> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Nhi V. Vo. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Last updated: November 18, 2025. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?c15de51d4bb57887caa2c21988d97279"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-502Y4YLYN2"></script> <script defer src="/assets/js/google-analytics-setup.js"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> </body> </html>