---
layout: default
title: Home
permalink: /
nav_order: 1
---
## **About Me** 
<div class="code-example fs-3 fw-400 lh-0.3" markdown="1">
<img src="assets/img/about_me_img/Screenshot (298).png" alt="Avatar of Nhi Vo" width="100" height="100" style="float: right; margin-left: 1rem;" /> 
Bioinformatics Research Associate with experience in developing and automating computational analysis of complex microbial data. 
</div>

## **Experience**
<div class="code-example fs-3 fw-400 lh-0.2" markdown="1">
**Research Support Associate 2**, *The Chisholm Lab*  
Massachusetts Institute of Technology, *Cambridge MA*  
Feb. 2023 - Present   
  
**Research Assistant**, *The Kawahara Lab*     
Florida Museum of Natural History, *Gainesville FL*   
May. 2022 - Aug. 2022     
 
**Data Analyst**, *Fernandez-Juricic Lab*    
Purdue University, *West Lafayette IN*   
May. 2022 - Aug. 2022   
  
**Research Assistant**, *Kihara Bioinformatics Laboratory*  
Purdue University, *West Lafayette IN*  
Aug. 2021 - May. 2022   
</div>

## **Education** 
<div class="code-example fs-3 fw-400 lh-0.3" markdown="1">
<strong>Bachelor of Science in Genetics</strong>  
Purdue University, *West Lafayette IN*  
Aug. 2019 - Dec. 2022  

<div class="fs-2"  markdown="1">
- *Minors*: Bioinformatics, Chemistry     
- GPA: 3.88  
</div>
</div>

## **Technical Skills**
<div class="code-example fs-2 fw-400 lh-0.3" markdown="1">
**Programming:** Python, Bash, Linux/Unix, SLURM/HPC, Git/GitHub, R  

**Workflow Languages**: Snakemake, Nextflow   

**Data Analysis & Visualization:** Pandas, NumPy, SciPy, Matplotlib, Plotly, Seaborn, Adobe Illustrator, Scikit-Learn   

**NGS Datasets:**   
- Illumina, Oxford Nanopore, PacBio HiFi, Sanger  
- RNA-seq/metatranscriptomics, DNA-seq/metagenomics, epigenomics  

**Bioinformatics Pipelines:**   
- Genome Assembly/Metagenomic Assembly (MAGs)  
- Comparative Genomics (Pangenomics) Analysis   
- Variant Calling: SNP, structural variant   
- Differential Gene Expression Analysis    
- Taxonomic Profiling   

**Bioinformatics**:    
- NGS Quality Control: bbduk (BBMap), trimmomatic, filtlong  
- Gene Annotation: Anvio, Prodigal, Prokka, eggNOG-mapper, COG, KEGG, GO  
- Read Mapping: bowtie2, BWA-MEM, minimap2  
- Genome Assembly:   
    - Assemblers: SPAdes, megahit, Flye/metaFlye, canu  
    - Quality: checkM, checkM2, GTDB-tk  
- Homology (sequence comparison): blastp, blastn, DIAMOND blast, mmseqs2  
- ANI: fastANI, skani  
- Files: BAM/SAM, VCF, pileup, GFF, FNA/FAA/FASTA, FASTQ  
- and more  
</div>

