---
layout: default
title: Home
permalink: /
nav_order: 1
---
![Avatar of Nhi Vo](assets/img/about_me_img/Screenshot (298).png){: width="100" height="100" .float-right .mx-4}   

## **About**  
<div class="code-example fs-3 fw-400 lh-0" markdown="1">
Bioinformatics Research Associate with experience in developing and automating computational analysis of complex microbial data. 
</div>

## **Experience**
<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**Bioinformatics Research Associate**  
Massachusetts Institute of Technology | The Chisholm Lab   
Cambridge MA | Feb 2023 - Present  

- Developed and automated a pipeline for bacterial genome assembly, binning, classification, and gene annotation for Illumina, Oxford Nanopore, and PacBio Hifi raw reads, designed for High-Performance Computing (HPC) clusters.
    - <a href="https://github.com/nhinvo/genome-assembly-pipeline" target="_blank">Genome Assembly Pipeline Github Repository</a>
- Optimized a bioinformatics pipeline for taxonomic classification and abundance analysis of marine picocyanobacteria in metagenomics samples.
    - <a href="https://github.com/jamesm224/ProSynTax-workflow/tree/main" target="_blank">ProSynTax Workflow Github Repository</a>
    - <a href="https://doi.org/10.1101/2025.03.20.644373" target="_blank">Pre-print</a>
- Developed bioinformatic workflows for methylation detection of PacBio (Revio, RSII, Sequel II, Sequel IIe) and Nanopore data.
    - <a href="https://github.com/nhinvo/pacbio-methylation" target="_blank">PacBio Methylation Detection Github Repository</a>
    - <a href="https://github.com/nhinvo/methylation-pipeline" target="_blank">Nanopore Methylation Detection Pipeline Github Repository</a>
- Performed Variant Calling analysis of microbial populations to detect strain-level genetic differences. 
    - <a href="https://github.com/nhinvo/biofilm-prochlorococcus" target="_blank">Variant Calling Analysis Github Repository</a>
- Performed Differential Gene Expression analysis to detect transcriptional responses of cyanobacterial populations to nutrient perturbations. 
    - <a href="https://doi.org/10.1093/ismeco/ycae131" target="_blank">Publication</a>
- Performed Bioinformatics analyses: BLAST, comparative genomics, ANI, Multiple Sequence Alignment.
</div>

<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**Research Assistant**  
Florida Museum of Natural History - The Kawahara Lab     
May 2022 - Aug 2022     
Gainesville FL  
- Ensured version control and reproducibility for phylogenetic analysis of 129 whole-genomes using Git and Conda
- Troubleshot errors and enhanced a phylogenetic analysis pipeline by integrating Python and Bash functionalities
</div>

<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**Technology Supervisor**   
Information Technology at Purdue  
Aug 2021 - Dec 2022   
West Lafayette IN  
- Supervised a team of 10 lab support personnel, ensuring timely and effective resolution of IT-related concerns. 
- Maintained communication and coordinated with various IT departments to resolve customer issues efficiently.
</div>

<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**Data Analyst**  
Purdue University  
May 2022 - Aug 2022   
West Lafayette IN  
- Processed experimental data for machine learning model training by filtering, parsing, and formatting using Python
- Created visualizations of biological data using Python plotting libraries and Adobe Illustrator, ranging from interactive
plots for data exploration to publication-ready figures
</div>

<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**Research Assistant**
Purdue University  
Aug 2021 - May 2022  
West Lafayette IN    
- Accelerated molecular docking simulations for 4,000+ protein-ligand pairs by 65% through leveraging parallelization and SLURM to optimize High-Performance Computing (HPC) utilization.
- Documented analysis workflows, detailing setup, execution, and debugging to ensure clarity and reproducibility. 
- Developed a cheminformatics workflow to evaluate the validity and accessibility of AI-generated drug molecules.
</div>

## **Technical Skills**
<div class="code-example fs-3 fw-400 lh-0" markdown="1">
- **Programming:** Python, Bash, Linux/Unix, SLURM/HPC, Git/GitHub, R
- **Workflow Languages**: Snakemake, Nextflow 
- **Data Analysis & Visualization:** Pandas, NumPy, SciPy, Matplotlib, Plotly, Seaborn, Adobe Illustrator, Scikit-Learn  
- **NGS Datasets:** 
    - Illumina, Oxford Nanopore, PacBio HiFi, Sanger
    - RNA-seq/metatranscriptomics, DNA-seq/metagenomics, epigenomics
- **Bioinformatics Pipelines:** 
    - Genome Assembly/Metagenomic Assembly (MAGs)
    - Comparative Genomics (Pangenomics) Analysis 
    - Variant Calling: SNP, structural variant 
    - Differential Gene Expression Analysis  
    - Taxonomic Profiling 
- **Bioinformatics**:  
    - NGS Quality Control: bbduk (BBMap), trimmomatic, filtlong
    - Gene Annotation: Anvio, Prodigal, Prokka, eggNOG-mapper, COG, KEGG, GO
    - Read Mapping: bowtie2, BWA-MEM, minimap2
    - Genome Assembly: 
        - Assemblers: SPAdes, megahit, Flye/metaFlye, canu
        - Quality: checkM, checkM2, GTDB-tk
    - Homology (sequence comparison): blastp, blastn, DIAMOND blast, mmseqs2
    - ANI: fastANI, skani
    - Files: BAM/SAM, VCF, pileup, GFF, FNA/FAA/FASTA, FASTQ
    - and more
</div>

## **Education** 
<div class="code-example fs-3 fw-400 lh-0" markdown="1">
**B.S. in Genetics**  
Purdue University  
Aug 2019 - Dec 2022  
West Lafayette IN  
- *Minors*: Bioinformatics, Chemistry     
- GPA: 3.88  
</div>

